llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+			{
+				GenericValue V = getOperandValue(I.getOperand(0));
+				if (V.IntVal == 0u)
+				{
+					if (isValueComputed(I.getOperand(2)))
+						return true;
+				}
+				else
+				{
+					if (isValueComputed(I.getOperand(1)))
+						return true;
+				}
+
+			}	
+		}
+		for (auto& op : I.operands())
+		{
+			if (!isValueComputed(op))
+				return false;
+		}
+		if (const SwitchInst* SI = dyn_cast <SwitchInst>(&I))
+		{
+			if ((getBitMask(SI->getCondition()) != BitMask::ALL) )
+				return false;
+		}
+		if (const BranchInst* SI = dyn_cast <BranchInst>(&I))
+		{
+			if (SI->isConditional())
+				if ((getBitMask(SI->getCondition()) != BitMask::ALL) )
+					return false;
+		}
+		return true;
+	}
+	bool hasToBeSkipped(const llvm::Instruction& I)
+	{
+		if (!areOperandsComputed(I))
+			return true;
+
+		switch (I.getOpcode())
+		{
+			case Instruction::LandingPad:
+			case Instruction::Invoke:
+			case Instruction::ExtractElement:
+			case Instruction::CatchPad:
+			case Instruction::CleanupPad:
+			case Instruction::CleanupRet:
+			case Instruction::Fence:
+			case Instruction::Unreachable:
+			case Instruction::InsertValue:
+			case Instruction::InsertElement:
+			case Instruction::Store:
+			case Instruction::VAArg:
+			{
+				return true;
+			}
+			case Instruction::Call:
+			{
+				const CallBase& CB = cast<CallBase>(I);
+				if (const Function* FF = dyn_cast<Function>(CB.getCalledFunction()))
+				{
+					//Recursive calls might be handled, but require the mapping to be tied to the ExecutionContext
+					if (isAlreadyInCallStack(FF))
+						break;
+
+					if (FF->isDeclaration())
+						break;
+
+					uint32_t argIndex=0;
+					for (auto& op : CB.args())
+					{
+						if (argIndex >= FF->getFunctionType()->getNumParams())
+						{
+							//It means it's a VAArg, break from the inner loop
+							break;
+						}
+						assignToMapsIfComputed(FF->getArg(argIndex), op);
+						argIndex++;
+					}
+					return false;
+				}
+				break;
+			}
+			case Instruction::Load:
+			{
+				const LoadInst& LI = cast<LoadInst>(I);
+
+				GenericValue SRC = getOperandValue(LI.getPointerOperand());
+				GenericValue *Ptr = (GenericValue*)GVTORP(SRC);
+
+				for (const auto& interval : validLoadIntervals)
+				{
+					if ((long long)Ptr >= interval.first && (long long)Ptr < interval.second)
+						return false;
+				}
+
+				// TODO(carlo): Possibly even Alloca might be proven valid
+				break;
+			}
+			default:
+			{
+				// TODO(carlo): review PtrToInt or IntToPtr
+				// TODO(carlo): actually the proper way would be have the default meaning it has to be skipped, so that the conservative answer is failure
+				return false;
+			}
+		}
+		return true;
+	}
+	void computeValidLoadIntervals(const Module& _module)
+	{
+		for (const auto& global : _module.globals())
+		{
+			const GlobalVariable* GV = dyn_cast<GlobalVariable>(&global);
+			if (!GV)
+				continue;
+			if (!GV->hasInitializer())
+				continue;
+			if (GV->isExternallyInitialized())
```

Comment:
Please review this usage for isExternallyInitialized. We check for internal linkage anway. How can an internal variable be externally initialized?

Reply:
I removed the condition (putting an assertion), since I am not actually completely sure of the relations between those + the fact that hasInternalLinkage might get relaxed at some point